### PR TITLE
ast/compile: 'every' rewriting steps

### PR DESCRIPTION
**Pull request by srenatus**
This PR covers the compiler doing it's job on `every`:

- resolve refs
- rewriting other declared vars in _domain_
- rewriting _its declared vars_ and others in _its body_
- rewriting dynamics in the body
- expanding expressions
- safety body reordering and var checking

### CODE REVIEW

**Comment by srenatus** [comment]
there's a broken test I've got to take care of (not sure yet if the test is bogus or the code). moving to draft.

File: ast/policy.go [hunk 1]
```diff
@@ -1491,6 +1491,21 @@ func (q *Every) Compare(other *Every) int {
...
+	r := NewVarSet()
+	if v := q.Value.Value.(Var); !v.IsWildcard() {
+		r.Add(v)
+	}
+	if q.Key != nil {
+		if v := q.Key.Value.(Var); !v.IsWildcard() {
+			r.Add(v)
+		}
+	}
+	return r
```
Comment (tsandall) [inline ast/policy.go:1506]: nit: should we use the VarVisitor here? What if someone writes `every k, [x, "foo"] in ...`

File: ast/compile.go [hunk 2]
```diff
@@ -3036,7 +3036,16 @@ func outputVarsForExpr(expr *Expr, arity func(Ref) int, safe VarSet) VarSet {
...
-		return outputVarsForTerms(terms.Domain, safe)
+		s := outputVarsForTerms(terms.Domain, safe)
+
+		cpy := safe.Copy()
+		if terms.Key != nil {
+			cpy.Add(terms.Key.Value.(Var))
+		}
+		cpy.Add(terms.Value.Value.(Var))
+
+		s.Update(outputVarsForBody(terms.Body, arity, cpy))
```
Comment (tsandall) [inline ast/compile.go:3059]: I'm surprised that we're adding the output vars from the body of the Every to the set returned here. Won't that make vars in the body of the every appear safe in the outside body? E.g.,

```
every x in y { z = x }
z > 10  # won't z be safe here? shouldn't it be unsafe?
```

File: ast/policy.go [hunk 3]
```diff
@@ -1479,6 +1491,21 @@ func (q *Every) Compare(other *Every) int {
 	return q.Body.Compare(other.Body)
 }
 
+// Vars returns the key and val arguments of an every expression,
+// if they are non-nil and not wildcards.
+func (q *Every) Vars() VarSet {
```
Comment (tsandall) [inline ast/policy.go:1496]: nit: could we call this something else? Perhaps `KeyValueVars()`. The reason is that we have `Vars()` functions on many other types and normally they either return ALL vars under the AST node or they allow callers to supply a parameter struct to customize it. This function diverges a bit by only looking at key/value.

File: ast/policy.go (hunk 1 above)
Comment (srenatus) [reply ast/policy.go:1506]: My thinking was that we'd not allow that, at least not right away. So in the first iteration, key and value need to be plain vars. 

Later on, we could relax that. WDYT?

File: ast/compile.go (hunk 2 above)
Comment (srenatus) [reply ast/compile.go:3059]: Yeah that's a good point. I don't know what I was thinking, but I'll dial that back. Thanks!

File: ast/compile.go [hunk 4]
```diff
@@ -2946,6 +2958,10 @@ func reorderBodyForClosures(arity func(Ref) int, globals VarSet, body Body) (Bod
 			vs := VarSet{}
 			WalkClosures(e, func(x interface{}) bool {
 				vis := &VarVisitor{vars: vs}
+				if ev, ok := x.(*Every); ok {
+					vis.Walk(ev.Body)
+					return true
+				}
```
Comment (srenatus) [inline ast/compile.go:2964]: This is a bit of a trap: `WalkClosures` checks that its node in question is of type *Every. So, when passing a function, you get an `x` of type *Every, and to walk its actual closure, the `every.Body`, it'll have to be done like this.

I wonder if there's a better approach here. Having `WalkClosures` apply `f` to `ev.Body` [here](https://github.com/open-policy-agent/opa/pull/4231/files#diff-0231d068c880c5fa17a560d427a97e6a8a2e4fc8c6ba048e4427baa7355b67ccR165) would not allow us to differentiate what we've been called on when we need it, [here](https://github.com/open-policy-agent/opa/pull/4231/files#diff-067a842cc5e1a348551f4c7656120105acedafb2faac6e7b7ff57b75c5349570R1466-R1468).

File: ast/compile.go [hunk 5]
```diff
@@ -1452,7 +1452,7 @@ func rewritePrintCalls(gen *localVarGenerator, getArity func(Ref) int, globals V
 	// Visit comprehension bodies recursively to ensure print statements inside
 	// those bodies only close over variables that are safe.
 	for i := range body {
-		if ContainsComprehensions(body[i]) {
+		if ContainsComprehensions(body[i]) || body[i].IsEvery() {
```
Comment (tsandall) [inline ast/compile.go:1455]: nit: What about adding a new "ContainsClosure" function that we can call instead to deal with comprehensions and every?

File: ast/compile.go (hunk 4 above)
Comment (tsandall) [reply ast/compile.go:2964]: I think this is unavoidable due to the scoping rules we've implemented for `every`. Presumably it'll be the same for any other keywords we add that support closures and local variable declaration. This seems like the best we can do.

File: ast/policy.go (hunk 1 above)
Comment (tsandall) [reply ast/policy.go:1506]: That makes sense, I'm fine with deferring the decision to support compound terms in the key-value position. That said, I could still see value in using the VarVisitor here assuming the check to prohibit compound terms is done during parsing. This way, if/when that limitaiton is lifted, this function will just do the right thing. It's completely your call though--I don't have a strong opinion.

**Comment by tsandall** [comment]
LGTM!

**Review by tsandall — DISMISSED** [review]

File: ast/compile.go (hunk 5 above)
Comment (srenatus) [reply ast/compile.go:1455]: ✔️ Good idea.

File: ast/policy.go (hunk 1 above)
Comment (srenatus) [reply ast/policy.go:1506]: Switched to VarVisitor. ✔️

**Pull request merged**